pdns/dnsdist-carbon.cc

```diff
@@ -31,255 +31,321 @@
+        for (const auto& item : v) {
+          str << base << name << "." << item.first << " " << item.second << " " << now << "\r\n";
+        }
+      }
+    }
 #endif /* HAVE_DNS_OVER_HTTPS */
 
-          {
-            std::string qname;
-            auto records = g_qcount.records.write_lock();
-            for (const auto &record : *records) {
-              qname = record.first;
-              boost::replace_all(qname, ".", "_");
-              str<<"dnsdist.querycount." << qname << ".queries " << record.second << " " << now << "\r\n";
-            }
-            records->clear();
-          }
+    {
+      std::string qname;
+      auto records = g_qcount.records.write_lock();
+      for (const auto& record : *records) {
+        qname = record.first;
+        boost::replace_all(qname, ".", "_");
+        str << "dnsdist.querycount." << qname << ".queries " << record.second << " " << now << "\r\n";
+      }
+      records->clear();
+    }
 
-          const string msg = str.str();
+    const string msg = str.str();
 
-          int ret = waitForRWData(s.getHandle(), false, 1 , 0);
-          if(ret <= 0 ) {
-            vinfolog("Unable to write data to carbon server on %s: %s", server.toStringWithPort(), (ret<0 ? stringerror() : "Timeout"));
-            continue;
-          }
-          s.setBlocking();
-          writen2(s.getHandle(), msg.c_str(), msg.size());
+    int ret = waitForRWData(s.getHandle(), false, 1, 0);
+    if (ret <= 0) {
+      vinfolog("Unable to write data to carbon server on %s: %s", server.toStringWithPort(), (ret < 0 ? stringerror() : "Timeout"));
+      return false;
+    }
+    s.setBlocking();
+    writen2(s.getHandle(), msg.c_str(), msg.size());
+  }
+  catch (const std::exception& e) {
+    warnlog("Problem sending carbon data to %s: %s", server.toStringWithPort(), e.what());
+    return false;
+  }
+
+  return true;
+}
+
+static void carbonHandler(Carbon::Endpoint endpoint)
+{
+  setThreadName("dnsdist/carbon");
+  const auto intervalUSec = endpoint.interval * 1000 * 1000;
+
+  try {
+    size_t consecutiveFailures = 0;
+    do {
+      DTime dt;
+      dt.set();
+      if (doOneCarbonExport(endpoint)) {
+        const auto elapsedUSec = dt.udiff();
+        if (elapsedUSec < 0 || static_cast<unsigned int>(elapsedUSec) <= intervalUSec) {
+          useconds_t toSleepUSec = intervalUSec - elapsedUSec;
+          usleep(toSleepUSec);
+        }
+        else {
+          vinfolog("Carbon export for %s took longer (%s usec) than the configured interval (%d usec)", endpoint.server.toStringWithPort(), elapsedUSec, intervalUSec);
         }
-        catch(const std::exception& e) {
-          warnlog("Problem sending carbon data: %s", e.what());
+        consecutiveFailures = 0;
+      }
+      else {
+        /* maximum interval between two attempts is 10 minutes */
+        const time_t maxBackOff = 10 * 60;
+        time_t backOff = 1;
+        double backOffTmp = std::pow(2.0, consecutiveFailures);
```

Comment:
likely some static analysis tool will complain about the implicit `size_t` -> `double` conversion.